frontend/src/features/permits/pages/Payment/PaymentRedirect.tsx

```diff
@@ -56,31 +63,45 @@ export const PaymentRedirect = () => {
   }, [paymentDetails.trnApproved]);
 
   useEffect(() => {
-
-    const applicationIds:string[] = [];
+    const applicationIds: string[] = [];
     if (transactionIdQuery?.isSuccess && issuedPermit?.current === false) {
-      
       transactionIdQuery?.data?.applicationDetails?.forEach((application) => {
         if (application?.applicationId) {
           applicationIds.push(application.applicationId);
         }
-      })
+      });
 
       if (paymentApproved === true) {
         // Payment successful, proceed to issue permit
         issuePermitsMutation.mutate(applicationIds);
         issuedPermit.current = true;
       } else if (paymentApproved === false) {
-        // Payment failed, redirect back to pay now page
-        navigate(SHOPPING_CART_ROUTES.DETAILS(true), {
-          replace: true,
-        });
+        // Add back to cart and then redirect to shopping cart.
+        if (!addToCartMutation.isPending && addToCartMutation.isIdle) {
+          addToCartMutation
+            .mutateAsync({
+              companyId: (companyId as number).toString(),
+              applicationIds,
+            })
+            .then(({ failure }) => {
+              // Cannot add back to cart
+              if (failure.length > 0) {
+                navigate(ERROR_ROUTES.UNEXPECTED);
```

Comment:
On a side note, would this (ie. adding back to cart failure) be considered an addition to the various payment failure (unhappy path) scenarios? I think the original plan (before introducing shopping cart), was to redirect to the application payment page (now cart page) regardless of what failures occurred during payment (with the query param ```paymentFailed=true```) so that an error banner will appear there (rather than to the unexpected error page). I understand that the user should now not be allowed to go back to shopping cart when adding back fails since the application wouldn't appear there, but does Michael know about this redirect to unexpected error page (and confirmed this as the current requirements going forward)?

Reply:
I'm yet to confirm the specifications on this one but I didn't want to leave it unattended. But yes, It's a new unhappy scenario as you've rightly pointed out.

Also, `WAITING_PAYMENT` items do show up in the AIP list. So, the user isn't fully losing track of it. They could try adding it back to the cart there too (once we add that feature, of course). Considering this, it should be okay for now, since there really isn't much else we can have the user do in case of errors in adding to cart.

Reply:
ok np, sounds good